## ORM

In [1]:
# OOP(객체지향프로그래밍) 이용한 DB 관리
# Object Relational mapping = Programming Technique



In [2]:
import sqlalchemy

In [3]:
#dialect+driver://username:password@host:port/database

# sqlite://<nohostname>/<path>
# where <path> is relative:
#engine = create_engine('sqlite:///foo.db')

sqlalchemy.__version__

'1.3.1'

In [4]:
from sqlalchemy import create_engine    #엔진 만들기

In [5]:
engine = create_engine("sqlite://", echo=True)
                       
#engine = create_engine("sqlite:///:memory:", echo=True)    #메모리 상에서 실행할 때
#engine = create_engine("sqlite:///test.db", echo=True)     #파일 상에서 실행할 때

In [6]:
print(engine)

Engine(sqlite://)


In [7]:
#Lazy connecting

In [8]:
engine

Engine(sqlite://)

In [28]:
from sqlalchemy import Table, Column, Integer, String, MetaData, ForeignKey

metadata = MetaData()
users = Table('users', metadata,
    Column('id',Integer, primary_key=True),
    Column('name',String),
    Column('fullname',String),
)

addresses = Table('addresses',metadata,
    Column('id',Integer, primary_key=True),
    Column('user_id',None, ForeignKey('users.id')),
    Column('email_address', String, nullable=False)
)

#metadata.create_all(engine)

In [29]:
metadata.tables

immutabledict({'users': Table('users', MetaData(bind=None), Column('id', Integer(), table=<users>, primary_key=True, nullable=False), Column('name', String(), table=<users>), Column('fullname', String(), table=<users>), schema=None), 'addresses': Table('addresses', MetaData(bind=None), Column('id', Integer(), table=<addresses>, primary_key=True, nullable=False), Column('user_id', Integer(), ForeignKey('users.id'), table=<addresses>), Column('email_address', String(), table=<addresses>, nullable=False), schema=None)})

In [15]:
metadata.bind #아무것도 없다

In [30]:
metadata.create_all(engine)  #engine에게 일해라 라고 시킨다(engine 연결)   #SQL문으로 바뀌는 과정

2019-07-09 11:40:15,652 INFO sqlalchemy.engine.base.Engine PRAGMA table_info("users")
2019-07-09 11:40:15,654 INFO sqlalchemy.engine.base.Engine ()
2019-07-09 11:40:15,655 INFO sqlalchemy.engine.base.Engine PRAGMA table_info("addresses")
2019-07-09 11:40:15,656 INFO sqlalchemy.engine.base.Engine ()


## INSERT

In [17]:
insert = users.insert()                  #SQL-INSERT 문을 만든다.
print(insert)

INSERT INTO users (id, name, fullname) VALUES (:id, :name, :fullname)


In [19]:
insert = users.insert().values(name='kim', fullname='Anonymous, Kim')      #SQL-INSERT 문의 내용을 만든다.
print(insert)

INSERT INTO users (name, fullname) VALUES (:name, :fullname)


In [20]:
insert.compile().params               #SQL-INSERT 문을 확정한다.(compile해서 parameter로 넘긴다)

{'name': 'kim', 'fullname': 'Anonymous, Kim'}

## Executing : Connection, execute, ResultProxy(Cursor)

In [21]:
conn = engine.connect()
conn

In [22]:
insert.bind = engine     #binding에 엔진을 달아준다. "일해라 노예야"
str(insert)

'INSERT INTO users (name, fullname) VALUES (?, ?)'

In [23]:
result = conn.execute(insert)

result.inserted_primary_key

2019-07-09 11:31:29,146 INFO sqlalchemy.engine.base.Engine INSERT INTO users (name, fullname) VALUES (?, ?)
2019-07-09 11:31:29,147 INFO sqlalchemy.engine.base.Engine ('kim', 'Anonymous, Kim')
2019-07-09 11:31:29,148 INFO sqlalchemy.engine.base.Engine COMMIT


[1]

### execute의 params 사용

In [24]:
insert = users.insert()
result = conn.execute(insert, name="lee", fullname="Unknown, Lee")     #데이터 삽입

result.inserted_primary_key              #2

2019-07-09 11:32:24,654 INFO sqlalchemy.engine.base.Engine INSERT INTO users (name, fullname) VALUES (?, ?)
2019-07-09 11:32:24,655 INFO sqlalchemy.engine.base.Engine ('lee', 'Unknown, Lee')
2019-07-09 11:32:24,656 INFO sqlalchemy.engine.base.Engine COMMIT


[2]

In [32]:
conn.execute(addresses.insert(), [{"user_id":1,"email_address":"anonyous.kim@test.com"},
                                    {"user_id":2,"email_address":"unknown.lee@test.com"}
                                 ])

2019-07-09 11:40:34,882 INFO sqlalchemy.engine.base.Engine INSERT INTO addresses (user_id, email_address) VALUES (?, ?)
2019-07-09 11:40:34,883 INFO sqlalchemy.engine.base.Engine ((1, 'anonyous.kim@test.com'), (2, 'unknown.lee@test.com'))
2019-07-09 11:40:34,884 INFO sqlalchemy.engine.base.Engine COMMIT


## SELECT

In [33]:
from sqlalchemy.sql import select

query = select([users])            # Table -> list
result = conn.execute(query)       # 

for row in result:
    print(row)

2019-07-09 11:42:26,160 INFO sqlalchemy.engine.base.Engine SELECT users.id, users.name, users.fullname 
FROM users
2019-07-09 11:42:26,161 INFO sqlalchemy.engine.base.Engine ()
(1, 'kim', 'Anonymous, Kim')
(2, 'lee', 'Unknown, Lee')


In [34]:
users  #Table

Table('users', MetaData(bind=None), Column('id', Integer(), table=<users>, primary_key=True, nullable=False), Column('name', String(), table=<users>), Column('fullname', String(), table=<users>), schema=None)

In [35]:
cur = conn.execute(select([users]))         #INSERT했던 내용 확인 가능.
cur.fetchall()                              #

2019-07-09 11:45:27,727 INFO sqlalchemy.engine.base.Engine SELECT users.id, users.name, users.fullname 
FROM users
2019-07-09 11:45:27,728 INFO sqlalchemy.engine.base.Engine ()


[(1, 'kim', 'Anonymous, Kim'), (2, 'lee', 'Unknown, Lee')]

## Conjunctions

In [36]:
from sqlalchemy import and_,or_,not_          #WHERE 절에 필요한 부분.

In [37]:
print(users.c.id == addresses.c.user_id)
print(users.c.id == 1)
print((users.c.id == 1).compile().params)
print(or_(users.c.id == addresses.c.user_id, users.c.id == 1))
print(and_(users.c.id == addresses.c.user_id, users.c.id == 1))
print(and_(or_(users.c.id == addresses.c.user_id, users.c.id == 1), addresses.c.email_address.like("a%")))
#print(((users.c.id == addresses.c.user_id) ))

users.id = addresses.user_id
users.id = :id_1
{'id_1': 1}
users.id = addresses.user_id OR users.id = :id_1
users.id = addresses.user_id AND users.id = :id_1
(users.id = addresses.user_id OR users.id = :id_1) AND addresses.email_address LIKE :email_address_1


## Selecting

In [38]:
result = conn.execute(select([users]).where(users.c.id==1))                 #from에 table 2개 넣으면 크로스 조인
for row in result:
    print(row)

2019-07-09 13:16:32,373 INFO sqlalchemy.engine.base.Engine SELECT users.id, users.name, users.fullname 
FROM users 
WHERE users.id = ?
2019-07-09 13:16:32,374 INFO sqlalchemy.engine.base.Engine (1,)
(1, 'kim', 'Anonymous, Kim')


In [40]:
result = conn.execute(select([users, addresses]).where(users.c.id==addresses.c.user_id))
# SELECT users, addresses FROM   WHERE users.id == addresses.user_id
for row in result:
    print(row)

2019-07-09 13:18:10,772 INFO sqlalchemy.engine.base.Engine SELECT users.id, users.name, users.fullname, addresses.id, addresses.user_id, addresses.email_address 
FROM users, addresses 
WHERE users.id = addresses.user_id
2019-07-09 13:18:10,773 INFO sqlalchemy.engine.base.Engine ()
(1, 'kim', 'Anonymous, Kim', 1, 1, 'anonyous.kim@test.com')
(2, 'lee', 'Unknown, Lee', 2, 2, 'unknown.lee@test.com')
(1, 'kim', 'Anonymous, Kim', 3, 1, 'anonyous.kim@test.com')
(2, 'lee', 'Unknown, Lee', 4, 2, 'unknown.lee@test.com')


## JOIN

In [44]:
from sqlalchemy import join

print(users.join(addresses))

print(users.join(addresses, users.c.id == addresses.c.user_id))               #동일.(cause PK, FK 지정했으니까.)

users JOIN addresses ON users.id = addresses.user_id
users JOIN addresses ON users.id = addresses.user_id


In [45]:
query = select([users.c.id, users.c.fullname, addresses.c.email_address])
              .select_from(users.join(addresses))         #JOIN 된 결과
    
    
# SELECT users.id, users.fullname, addresses.email_address 
# FROM users JOIN addresses ON users.id = addresses.user_id    
    

result = conn.execute(query).fetchall()

2019-07-09 13:24:25,982 INFO sqlalchemy.engine.base.Engine SELECT users.id, users.fullname, addresses.email_address 
FROM users JOIN addresses ON users.id = addresses.user_id
2019-07-09 13:24:25,984 INFO sqlalchemy.engine.base.Engine ()


In [46]:
for row in result:
    print(row)

(1, 'Anonymous, Kim', 'anonyous.kim@test.com')
(2, 'Unknown, Lee', 'unknown.lee@test.com')
(1, 'Anonymous, Kim', 'anonyous.kim@test.com')
(2, 'Unknown, Lee', 'unknown.lee@test.com')


# 새 시작

In [66]:
metadata.tables

immutabledict({})

In [67]:
metadata.clear()

In [68]:
engine

Engine(sqlite://)

In [71]:
artist = Table("Artist", metadata, 
            Column("id", Integer, primary_key=True),
            Column("name", String, nullable=False),
              extend_existing=True)

album = Table("Album", metadata, 
            Column("id", Integer, primary_key=True),
            Column("title", String, nullable=False),
              extend_existing=True)

genre = Table("Genre", metadata,
             Column("id", Integer, primary_key=True))

Track = Tablle("Track", metadata, 
               Column("id", Integer, primary_key=True),
               Column("title", Integer, nullable=True),
               Column("length", Integer, nullable=True),
               Column("rating", Integer, nullable=True),
               Column("count", Integer, nullable=True),
               Column("album_id", Integer, ForeignKey("Album.id")),
               Column("id", Integer, ForeignKey("Genre.id"))
               extend_existing=True)

metadata.create_all

SyntaxError: invalid syntax (<ipython-input-71-0dd4bb1e4900>, line 22)

## SHOW TABLES / INSERT

In [57]:
tables = metadata.tables
for table in tables:
    print(table)
    
for table in engine.table_names():
    print(table)

2019-07-09 13:44:08,354 INFO sqlalchemy.engine.base.Engine SELECT name FROM sqlite_master WHERE type='table' ORDER BY name
2019-07-09 13:44:08,355 INFO sqlalchemy.engine.base.Engine ()
addresses
users


In [59]:
conn.execute(artist.insert(), [{"name":"Led Zepplin"}, {"name":"AC/DC"}])

conn.execute(album.insert(), [{"title":"TV", "artist_id":1}, {"title":"Who made Who", "artist_id":2}])

conn.execute(genre.insert(), [{"name":"Rock"}, {"name":"Metal"}])

conn.execute(track.insert(), [{"title":"Black Dog", "rating":5, "length":297, "count":0, "album_id":1, "genre_id":1}, 
                              {"title":"Stairway", "rating":5, "length":482, "count":0, "album_id":1, "genre_id":1},
                             {"title":"About to rock", "rating":5, "length":313, "count":0, "album_id":2, "genre_id":2},
                             {"title":"Who Made Who", "rating":5, "length":297, "count":0, "album_id":2, "genre_id":2},])

NameError: name 'artist' is not defined

## WHERE

In [60]:
trackResult = conn.execute(select([track]).where(and_(track.c.album_id==1, track.c.genre_id==2,)))

# SELECT track FROM ___ WHERE track.album_id ==1 and track.genre_id = 2

for row in result:
    print(row)

NameError: name 'track' is not defined

In [65]:
# Multiple Join

print(track.join(album))
print(track.join(album).join(genre))
print(track.join(album).join(artist))
print(track.join(album).join(genre).join(artist))

result = conn.execute(select([track.c.title, album.c.title, genre.c.name, artist.c.name])
                      .select_from(track.join(album).join(genre).join(artist)))

for row in result.fetchall():
    print(row)
    
result = conn.execute(track.select().select_from(track.join(album).join(genre).join(artist))
                      .where(and_(genre.c.id==1, artist.c.id==1)))


NameError: name 'track' is not defined

## Open/Close

In [72]:
#close

conn.close()
metadata.clear()

In [73]:
metadata.tables

immutabledict({})

In [74]:
engine

Engine(sqlite://)

In [75]:
tables = metadata.tables
for table in tables:
    print(table)
    
for table in engine.table_names():
    print(table)

2019-07-09 14:21:41,163 INFO sqlalchemy.engine.base.Engine SELECT name FROM sqlite_master WHERE type='table' ORDER BY name
2019-07-09 14:21:41,164 INFO sqlalchemy.engine.base.Engine ()
addresses
users


In [76]:
# Open

metadata = MetaData(bind=engine, reflect=True)
metadata.reflect(bind=engine)

for row in metadata.tables:
    print(row)

2019-07-09 14:24:04,159 INFO sqlalchemy.engine.base.Engine SELECT name FROM sqlite_master WHERE type='table' ORDER BY name
2019-07-09 14:24:04,160 INFO sqlalchemy.engine.base.Engine ()
2019-07-09 14:24:04,162 INFO sqlalchemy.engine.base.Engine PRAGMA table_info("addresses")
2019-07-09 14:24:04,163 INFO sqlalchemy.engine.base.Engine ()
2019-07-09 14:24:04,166 INFO sqlalchemy.engine.base.Engine SELECT sql FROM  (SELECT * FROM sqlite_master UNION ALL   SELECT * FROM sqlite_temp_master) WHERE name = 'addresses' AND type = 'table'
2019-07-09 14:24:04,167 INFO sqlalchemy.engine.base.Engine ()
2019-07-09 14:24:04,170 INFO sqlalchemy.engine.base.Engine PRAGMA foreign_key_list("addresses")
2019-07-09 14:24:04,171 INFO sqlalchemy.engine.base.Engine ()
2019-07-09 14:24:04,173 INFO sqlalchemy.engine.base.Engine SELECT sql FROM  (SELECT * FROM sqlite_master UNION ALL   SELECT * FROM sqlite_temp_master) WHERE name = 'addresses' AND type = 'table'
2019-07-09 14:24:04,173 INFO sqlalchemy.engine.base.E

C:\Users\David\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SADeprecationWarning: The MetaData.reflect flag is deprecated and will be removed in a future release.   Please use the MetaData.reflect() method.
  This is separate from the ipykernel package so we can avoid doing imports until


## Declare

In [77]:
from sqlalchemy import create_engine
engine = create_engine("sqlite:///:memory:", echo=True)

In [78]:
from sqlalchemy.ext.declarative import declarative_base

base = declarative_base()

# Create

In [81]:
class User(base):
    __tablename__ = "users"
    
    id = Column(Integer, primary_key=True)
    name = Column(String)
    fullname = Column(String)
    password = Column("passwd", String)
    
    def __repr__(self):
        return "<T'User(name='%s', fullname='%s', password='%s'>" % (self.name, self.fullname, self.password)
    
    
# CREATE TABLE users (
# 	id INTEGER NOT NULL, 
# 	name VARCHAR, 
# 	fullname VARCHAR, 
# 	passwd VARCHAR, 
# 	PRIMARY KEY (id)

In [82]:
User.__table__

Table('users', MetaData(bind=None), Column('id', Integer(), table=<users>, primary_key=True, nullable=False), Column('name', String(), table=<users>), Column('fullname', String(), table=<users>), Column('passwd', String(), table=<users>), schema=None)

In [87]:
print(User.__mapper__)              #Mapping 여부 => 즉, User 클래스를 건드리면 users DB를 건드는것과 같다.

mapped class User->users


In [83]:
#Create Table
base.metadata.create_all(engine)

2019-07-09 14:32:35,028 INFO sqlalchemy.engine.base.Engine SELECT CAST('test plain returns' AS VARCHAR(60)) AS anon_1
2019-07-09 14:32:35,030 INFO sqlalchemy.engine.base.Engine ()
2019-07-09 14:32:35,031 INFO sqlalchemy.engine.base.Engine SELECT CAST('test unicode returns' AS VARCHAR(60)) AS anon_1
2019-07-09 14:32:35,031 INFO sqlalchemy.engine.base.Engine ()
2019-07-09 14:32:35,033 INFO sqlalchemy.engine.base.Engine PRAGMA table_info("users")
2019-07-09 14:32:35,033 INFO sqlalchemy.engine.base.Engine ()
2019-07-09 14:32:35,035 INFO sqlalchemy.engine.base.Engine 
CREATE TABLE users (
	id INTEGER NOT NULL, 
	name VARCHAR, 
	fullname VARCHAR, 
	passwd VARCHAR, 
	PRIMARY KEY (id)
)


2019-07-09 14:32:35,036 INFO sqlalchemy.engine.base.Engine ()
2019-07-09 14:32:35,038 INFO sqlalchemy.engine.base.Engine COMMIT


In [84]:
# Create Instance

kim = User(name="kim", fullname="anonymous, Kim", password="kimbap heaven")

print(kim)
print(kim.id)

<T'User(name='kim', fullname='anonymous, Kim', password='kimbap heaven'>
None


# Session

In [88]:
from sqlalchemy.orm import sessionmaker

Session = sessionmaker(bind=engine)
session = Session()

In [89]:
# add(instance, _warn=True)

session.add(kim)

In [90]:
#add_all(instance)

session.add_all([
    User(name="lee", fullname="unknown, Lee", password="123456789a"),
    User(name="park", fullname="nobody, Park", password="Parking in Park")
])

In [92]:
# Commit

for row in session.query(User):
    print(type(row))
    print(row.id, row.name, row.fullname, row.password)
    
    
# 1 kim anonymous, Kim kimbap heaven
# <class '__main__.User'>
# 2 lee unknown, Lee 123456789a
# <class '__main__.User'>
# 3 park nobody, Park Parking in Park

2019-07-09 14:47:14,398 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2019-07-09 14:47:14,401 INFO sqlalchemy.engine.base.Engine INSERT INTO users (name, fullname, passwd) VALUES (?, ?, ?)
2019-07-09 14:47:14,402 INFO sqlalchemy.engine.base.Engine ('kim', 'anonymous, Kim', 'kimbap heaven')
2019-07-09 14:47:14,404 INFO sqlalchemy.engine.base.Engine INSERT INTO users (name, fullname, passwd) VALUES (?, ?, ?)
2019-07-09 14:47:14,405 INFO sqlalchemy.engine.base.Engine ('lee', 'unknown, Lee', '123456789a')
2019-07-09 14:47:14,406 INFO sqlalchemy.engine.base.Engine INSERT INTO users (name, fullname, passwd) VALUES (?, ?, ?)
2019-07-09 14:47:14,407 INFO sqlalchemy.engine.base.Engine ('park', 'nobody, Park', 'Parking in Park')
2019-07-09 14:47:14,409 INFO sqlalchemy.engine.base.Engine SELECT users.passwd AS users_passwd, users.id AS users_id, users.name AS users_name, users.fullname AS users_fullname 
FROM users
2019-07-09 14:47:14,411 INFO sqlalchemy.engine.base.Engine ()
<class '__main

In [100]:
eo = User(name="eo", fullname="anonymous, eo", password="mcdonald")
session.add(eo)

eo.password = "password"
session.dirty

IdentitySet([<T'User(name='kim', fullname='anonymous, Kim', password='password'>])

In [101]:
session.is_modified(eo)

True

In [98]:
kim.password = "password"
session.dirty

IdentitySet([<T'User(name='kim', fullname='anonymous, Kim', password='password'>])

In [99]:
session.is_modified(kim)

True

In [102]:
print(kim.id)

1


In [104]:
for row in session.query(User):
    print(type(row))
    print(row.id, row.name, row.fullname, row.password)

2019-07-09 15:11:07,997 INFO sqlalchemy.engine.base.Engine UPDATE users SET passwd=? WHERE users.id = ?
2019-07-09 15:11:07,999 INFO sqlalchemy.engine.base.Engine ('password', 1)
2019-07-09 15:11:08,000 INFO sqlalchemy.engine.base.Engine INSERT INTO users (name, fullname, passwd) VALUES (?, ?, ?)
2019-07-09 15:11:08,001 INFO sqlalchemy.engine.base.Engine ('eo', 'anonymous, eo', 'password')
2019-07-09 15:11:08,002 INFO sqlalchemy.engine.base.Engine INSERT INTO users (name, fullname, passwd) VALUES (?, ?, ?)
2019-07-09 15:11:08,003 INFO sqlalchemy.engine.base.Engine ('eo', 'anonymous, eo', 'password')
2019-07-09 15:11:08,004 INFO sqlalchemy.engine.base.Engine INSERT INTO users (name, fullname, passwd) VALUES (?, ?, ?)
2019-07-09 15:11:08,005 INFO sqlalchemy.engine.base.Engine ('eo', 'anonymous, eo', 'password')
2019-07-09 15:11:08,006 INFO sqlalchemy.engine.base.Engine INSERT INTO users (name, fullname, passwd) VALUES (?, ?, ?)
2019-07-09 15:11:08,007 INFO sqlalchemy.engine.base.Engine (

In [105]:
# filter

for row in session.query(User.id, User.fullname).filter(User.name == "Lee"):
    print(type(row))
    print(row.id, row.fullname)

2019-07-09 15:12:29,320 INFO sqlalchemy.engine.base.Engine SELECT users.id AS users_id, users.fullname AS users_fullname 
FROM users 
WHERE users.name = ?
2019-07-09 15:12:29,321 INFO sqlalchemy.engine.base.Engine ('Lee',)


In [109]:
# filter_by

for row in session.query(User.id, User.fullname).filter_by(name = "lee"):
    print(type(row))
    print(row.id, row.fullname)

2019-07-09 15:13:24,444 INFO sqlalchemy.engine.base.Engine SELECT users.id AS users_id, users.fullname AS users_fullname 
FROM users 
WHERE users.name = ?
2019-07-09 15:13:24,446 INFO sqlalchemy.engine.base.Engine ('lee',)
<class 'sqlalchemy.util._collections.result'>
2 unknown, Lee


# 적용 : album

In [117]:
metadata.clear()

In [118]:
metadata.tables

immutabledict({})

In [119]:
from sqlalchemy import create_engine
engine = create_engine("sqlite:///:memory:", echo=True)

from sqlalchemy.ext.declarative import declarative_base

base = declarative_base()

In [120]:
class Artist(base):
    __tablename__ = "Artist"
    
    id = Column(Integer, primary_key=True)
    name = Column(String)
    
class Album(base):
    __tablename__ = "Album"
    
    id = Column(Integer, primary_key=True)
    title = Column(String)
    artist_id = Column(Integer, ForeignKey("Artist.id"))
    
class Genre(base):
    __tablename__ = "Genre"
    
    id = Column(Integer, primary_key=True)
    name = Column(String)
    
class Track(base):
    __tablename__ = "Track"
    
    id = Column(Integer, primary_key=True)
    title = Column(String)
    length = Column(Integer)
    rating = Column(Integer)
    count = Column(Integer)
    album_id = Column(Integer, ForeignKey("Album.id"))
    genre_id = Column(Integer, ForeignKey("Genre.id"))

In [126]:
artist1 = Artist(name = "Led Zepplin")
artist2 = Artist(name = "AC/DC")

session.add_all([artist1, artist2])
session.commit()

album = [Album(title="IV", artist_id = artist1.id),
         Album(title="Who Made Who", artist_id = artist2.id)]

session.add_all(album)
session.commit()

session.add_all([Genre(name="Rock"), Genre(name="Metal")])
session.commit()

album1 = session.query(Album).filter(Album.artist_id==artist1.id).one()
album2 = session.query(Album).filter(Album.artist_id==artist2.id).one()

genre1 = session.query(Genre).filter(Genre.name=="Rock").filter(Genre.id==1).one()
genre2 = session.query(Genre).filter(Genre.name=="Metal").filter(Genre.id==2).one()

track = [Track(title="Black Dog", rating=5, length=297, count=0, album_id=album1.id, genre_id=genree1.id)]

InvalidRequestError: This Session's transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: (sqlite3.OperationalError) no such table: Genre
[SQL: INSERT INTO "Genre" (name) VALUES (?)]
[parameters: ('Rock',)]
(Background on this error at: http://sqlalche.me/e/e3q8)

# Relationship

In [128]:
metadata.clear()

In [129]:
metadata.tables

immutabledict({})

In [132]:
from sqlalchemy import create_engine
engine = create_engine("sqlite:///:memory:", echo=True)

from sqlalchemy.ext.declarative import declarative_base

base = declarative_base()

from sqlalchemy.orm import relationship

In [133]:
class Artist(base):
    __tablename__ = "Artist"
    
    id = Column(Integer, primary_key=True)
    name = Column(String)

    albumList = relationship("Album", back_populates="artist")
    
class Album(base):
    __tablename__ = "Album"
    
    id = Column(Integer, primary_key=True)
    title = Column(String)
    artist_id = Column(Integer, ForeignKey("Artist.id"))
    
    artist = relationship("Artist", back_populates="albumList", uselist=False)
    trackList = relationship("Track", back_populates="album")
    
class Genre(base):
    __tablename__ = "Genre"
    
    id = Column(Integer, primary_key=True)
    name = Column(String)
    
    trackList = relationship("Track", back_populates="genre")
    
class Track(base):
    __tablename__ = "Track"
    
    id = Column(Integer, primary_key=True)
    title = Column(String)
    length = Column(Integer)
    rating = Column(Integer)
    count = Column(Integer)
    album_id = Column(Integer, ForeignKey("Album.id"))
    genre_id = Column(Integer, ForeignKey("Genre.id"))
    
    album=relationship("Album", back_populates="trackList", uselist=False)
    genre=relationship("Genre", back_populates="trackList", uselist=False)

### INSERT

In [136]:
track1 = Track(title="Black Dog", rating=5, length=297, count=0)
track2 = Track(title="Stairway", rating=5, length=482, count=0)
track3 = Track(title="About to rock", rating=5, length=313, count=0)
track4 = Track(title="Who Made Who", rating=5, length=297, count=0)

In [137]:
track1.album = track2.album = Album(title="IV")
track3.album = track4.album = Album(title="Who Made Who")

In [138]:
track1.genre = track2.genre = Genre(name="Rock")
track3.genre = track4.genre = Genre(name="Metal")

In [139]:
track1.album.artist = track2.album.artist = Artist(name="Led Zepplin")
track3.album.artist = track4.album.artist = Artist(name="AC/DC")

### SELECT

In [140]:
print("Title: %s, Album: %s, Genre: %s, Artist: %s" %
    (track1.title, track1.album.title, track1.genre.name, track1.album.artist.name))
print("Title: %s, Album: %s, Genre: %s, Artist: %s" %
    (track3.title, track3.album.title, track3.genre.name, track3.album.artist.name))

Title: Black Dog, Album: IV, Genre: Rock, Artist: Led Zepplin
Title: About to rock, Album: Who Made Who, Genre: Metal, Artist: AC/DC


In [142]:
print("TrackID: %d, AlbumID: %d, GenreID: %d, Artist: %d" %
     (track1.id, track1.album.id, track1.genre.id, track1.album.artist.id))

print("TrackID: %d, AlbumID: %d, GenreID: %d, Artist: %d" %
     (track3.id, track3.album.id, track3.genre.id, track3.album.artist.id))

TypeError: %d format: a number is required, not NoneType

# Regular Expression (RE)

## => 주어진 텍스트로부터 structure를 가지게끔 하려는 목적. (탐색, replacing, parsing)

In [147]:
# 속도가 빠름 빠름 빠름(매우 많은 텍스트 데이터 처리)
# String searching / input validation(인풋 형식 체크)
# 단점 : learning curve(진입장벽 있음. 익숙해지는데 시간걸림)

# Anchor : ^(시작) $(끝)

.(dot) : 표현할 수 있는 모든 글자(하나)
a.b  -> a+모든문자+b -> aab,a0b,apb,

*(astersk) : 반복.
ab*c -> a+b(0번이상반복) + c
->  ac, abc, abbbbbbc,

+(plus) 1번이상반복
ab+c -> a+b(1번이상반복)+c
-> ac, abc, abbbbbc

{n|min, |min, max}
ab{1}c, ab{2,6}c
-> a+b(1번반복)+c / a+b(2~6번 반복) +c

?(question) : boolean.
ab?c, b{0,1}
-> a+b (1개 있거나,없거나)+c


^x : 문자열이 x로 시작.
x$ : 문자열이 x로 끝.
.x : 뭐시기뭐시기x             ★
x+ : x가 1번이상 반복
x? : x가 존재하던가/존재하지 않던가   ★
x|y : x이거나/y이거나
(x) : 그룹화
(x)(y) : 그룹1, 그룹2
(x)(?:y) : (나중에)
x(n) : x를 n번 반복한 문자를 찾는다


[](square braket) 
[abc] : a or b or c
[a-z] : a부터 z까지
[^a-z] : a부터 z까지를 제외 (NOT의 의미) (숫자, 대소문자 포함)

[xy] : x or y
[^xy] : not x, y
[x-z] : x에서 z까지
\ : 용법이 다름.
\w : 알파벳+숫자+_
\W : (알파벳+숫자+_) 를 제외한 모든 문자
\s : 공백
\S : 공백이 아닌 문자


Greedy vs Lazy (Greedy 방식에 ? 붙임)
Greedy : 마지막 가능 매치
Lazy : 처음 가능 매치

(https://www.regexr.com)
cat|mat : cat or mat

gr(e|a)y : grey or gray

py(pi|thon(n|nic) : pypi / python / pythonic

^obje : object or object-oriented
^2018 : 2018 or 2018-07-10
gram$ : program or kilogram

bat. : bat or bats or bata
[0123456789] : any digit
[a-zA-Z] 모든 알파벳.(대소문자 모두)
[^aeiou] : 모음이 아닌것.

### 도메인 거르기
\b[\w.%+-]+@[\w.-]+\.[a-zA-Z]{2,6}\b(|\.[a-zA-z]{2,3}\b)

### HTML 태그 거르기
<([a-zA-Z][A-Z0-9]*)[^>]*>(.*?)<\/\1>         (\1 : 1번째 그룹)
             <h1 style="">내용</h1>
             
### 전화번호 거르기
((\+?)([0-9]{0,2})(\-|\s)([0-9]{0,2})(\-|\s))?([0-9]{2,3})(\-|\s)([0-9]{3,4})(\-|\s)([0-9]{4})             
         

In [148]:
#파이썬에서의 Regular Expression
import re

In [149]:
# re.compile(pattern, flags=0)
# re.search(pattern, string, flags=0)   중간에 있는 글자 찾기
# re.match(pattern, string, flags=0)    주어진 스트링을 처음부터

# re.split(pattern, string, maxsplit=0, flags=0)
# findall() : 정규식과 매치되는 모든 문자열을 리스트로 리턴

cont = "Hello World"
print(re.search("W",cont))
print(re.match("W",cont))

<re.Match object; span=(6, 7), match='W'>
None


In [150]:
# re.sub() 대체시키는 함수.
data = """
park 800905-1049118
kim  700905-1059119
"""

result = []
for line in data.split("\n"):
    word_result=[]
    for word in line.split(" "):
        if len(word) == 14 and word[:6].isdigit() and word[7:].isdigit():
            word = word[:6] + "-" + "*******"
        word_result.append(word)
    result.append(" ".join(word_result))
print("\n".join(result))


park 800905-*******
kim  700905-*******



In [151]:
data = """
park 800905-1049118
kim  700905-1059119
"""

pat = re.compile("(\d{6})[-]\d{7}")
print(pat.sub("\g<1>-*******",data))


park 800905-*******
kim  700905-*******



In [155]:
p = re.compile('Crow|Servo')
m = p.match('CrowHello')
mm = p.match('ServoBye')
print(m,"\n",mm)

<re.Match object; span=(0, 4), match='Crow'> 
 <re.Match object; span=(0, 5), match='Servo'>


In [156]:
def res(m,n):
    return re.search(m,n)

In [158]:
print(res('^Life', 'Life is too short'))
print(res('^Life', 'My Life'))

<re.Match object; span=(0, 4), match='Life'>
None


In [159]:
# \b : word boundary
p=re.compile(r'\bclass\b')
print(p.search('no class at all'))
print(p.search('one subclass is'))
print(p.search('the declassified algorithm'))

<re.Match object; span=(3, 8), match='class'>
None
None


In [160]:
p=re.compile(r'\Bclass\B')
print(p.search('no class at all'))
print(p.search('one subclass is'))
print(p.search('the declassified algorithm'))

None
None
<re.Match object; span=(6, 11), match='class'>


In [162]:
p=re.compile(r'\Bclass\b')
print(p.search('no class at all'))
print(p.search('one subclass is'))
print(p.search('the declassified algorithm'))

None
<re.Match object; span=(7, 12), match='class'>
None


In [163]:
p = re.compile(r"\w+\s+\d+[-]\d+[-]\d+")
m = p.search("park 010-1234-1234")
print(m)

<re.Match object; span=(0, 18), match='park 010-1234-1234'>


In [164]:
p = re.compile(r"(\w+)\s+\d+[-]\d+[-]\d+")
m = p.search("park 010-1234-1234")
print(m.group(1))

park


## Group

In [168]:
m = re.match(r"(\w+) (\w+)", "Isaac Newton, physicist")
print(m.group(0))  #그룹 전체
print(m.group(1))  #그룹 1
print(m.group(2))  #그룹 2
print(m.group(1,2))

Isaac Newton
Isaac
Newton
('Isaac', 'Newton')


In [10]:
p = re.compile(r"(\w+) (\w+)")
m = p.search("Isaac Newton, physicist")
p.sub("\g<2> \g<1>", "Isaac Newton, physicist")

'Newton Isaac, physicist'

In [169]:
# 숙제

In [2]:
import re
p = re.compile("[a-z]+")
m = p.search("5 python")
m.start() + m.end()      

10

In [3]:
m.start()

2

In [4]:
m.end()

8

In [15]:
data = """
park 010-9999-9988
kim 010-9909-7789
lee 010-8789-7768
"""
p = re.compile("(\d{3})[-](\d{4})[-](\d{4})")  #g<1> g<2> g<3>
print(p.sub("\g<1>-\g<2>-****", data))


park 010-9999-****
kim 010-9909-****
lee 010-8789-****

